moquette-0.16/broker/src/main/java/io/moquette/broker/Session.java

```diff
@@ -0,0 +1,494 @@
+        return status.get() == SessionStatus.DISCONNECTED;
+    }
+
+    public boolean connected() {
+        return status.get() == SessionStatus.CONNECTED;
+    }
+
+    public String getClientID() {
+        return clientId;
+    }
+
+    public List<Subscription> getSubscriptions() {
+        return new ArrayList<>(subscriptions);
+    }
+
+    public void addSubscriptions(List<Subscription> newSubscriptions) {
+        subscriptions.addAll(newSubscriptions);
+    }
+
+    public void removeSubscription(Topic topic) {
+        subscriptions.remove(new Subscription(clientId, topic, MqttQoS.EXACTLY_ONCE));
+    }
+
+    public boolean hasWill() {
+        return will != null;
+    }
+
+    public Will getWill() {
+        return will;
+    }
+
+    boolean assignState(SessionStatus expected, SessionStatus newState) {
+        return status.compareAndSet(expected, newState);
+    }
+
+    public void closeImmediately() {
+        mqttConnection.dropConnection();
+        mqttConnection = null;
+        status.set(SessionStatus.DISCONNECTED);
+    }
+
+    public void disconnect() {
+        final boolean res = assignState(SessionStatus.CONNECTED, SessionStatus.DISCONNECTING);
+        if (!res) {
+            // someone already moved away from CONNECTED
+            // TODO what to do?
+            return;
+        }
+
+        mqttConnection = null;
+        will = null;
+
+        assignState(SessionStatus.DISCONNECTING, SessionStatus.DISCONNECTED);
+    }
+
+    boolean isClean() {
+        return clean;
+    }
+
+    public void processPubRec(int pubRecPacketId) {
+        // Message discarded, make sure any buffers in it are released
+        SessionRegistry.EnqueuedMessage removed = inflightWindow.remove(pubRecPacketId);
+        if (removed == null) {
+            LOG.warn("Received a PUBREC with not matching packetId");
+            return;
+        }
+        removed.release();
+        if (removed instanceof SessionRegistry.PubRelMarker) {
+            LOG.info("Received a PUBREC for packetId that was already moved in second step of Qos2");
+            return;
+        }
+
+        if (mqttConnection == null) {
+            return;
+        }
+        inflightWindow.put(pubRecPacketId, new SessionRegistry.PubRelMarker());
+        inflightTimeouts.add(new InFlightPacket(pubRecPacketId, FLIGHT_BEFORE_RESEND_MS));
+        MqttMessage pubRel = MQTTConnection.pubrel(pubRecPacketId);
+        mqttConnection.sendIfWritableElseDrop(pubRel);
+
+        drainQueueToConnection();
+    }
+
+    public void processPubComp(int messageID) {
+        // Message discarded, make sure any buffers in it are released
+        SessionRegistry.EnqueuedMessage removed = inflightWindow.remove(messageID);
+        if (removed == null) {
+            LOG.warn("Received a PUBCOMP with not matching packetId");
+            return;
+        }
+        removed.release();
+        inflightSlots.incrementAndGet();
+        drainQueueToConnection();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 223:230, 348:356.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.